docs/EpochBasedMemoryManagement.md

```diff
@@ -0,0 +1,129 @@
+# Epoch based memory management.
+
+## Overview
+
+The eBPF for Windows project uses an epoch based scheme for managing
+memory that permits a certain class of lock free operations,
+specifically the ability to implement lock free hash tables and other
+structures that require "read copy update" aka RCU semantics.
+
+Epoch driven memory management is an area that has been covered extensively by
+academic papers (as an example [Interval-Based Memory Reclamation
+(rochester.edu)](https://www.cs.rochester.edu/~scott/papers/2018_PPoPP_IBR.pdf)).
+The approach taken in this project is a simplification of several
+different approaches outlined in various research papers with the result
+being a tradeoff between performance and code complexity.
+
+In the context of this project's epoch memory management module
+(referred to as epoch module here in), the term epoch is intended to
+mean a period of indeterminate length. At the heart of the epoch module
+are two clocks:
+
+1)  _ebpf_current_epoch
+
+2)  _ebpf_release_epoch
+
+The first clock (_ebpf_current_epoch) tracks the current "time" in the
+system, with this being a clock that monotonically increases. The second clock
+(_ebpf_release_epoch) tracks the highest epoch that no longer has any
+code executing in it.
+
+Every execution context (a thread at passive IRQL or a DPC running at
+dispatch IRQL) is associated with the point in time when execution began
+(i.e. the value of the _ebpf_current_epoch clock at the point where it
+began execution). All memory that the execution context could touch
+during its execution is part of that epoch.
+
+When memory is no longer needed, it is first made non-reachable (all
+pointers to it are removed) after which it is stamped with the current
+epoch and inserted into a "free list". The timestamp is point in time
```

Comment:
```suggestion
epoch and inserted into a "free list". The timestamp is the point in time
```
also I note that it is "timestamp" here, "time stamp" in line 42, and "stamp" in 77.  would be nice to be consistent.